nucypher/acumen/perception.py

```diff
@@ -168,8 +187,82 @@ def abridged_state_details(state):
+            return iter(self)
+        if label not in BUCKETS:
+            raise self.UnknownLabel(f'{label} is not a valid node label')
+        try:
+            nodes = self.__marked[label]
+        except KeyError:
+            return iter(list())  # empty
+        return iter(nodes)
+
+    def get_label(self, node: "Teacher") -> Union["Teacher", None]:
+        for _label in BUCKETS:
+            if node in self.__marked[_label]:
+                return _label
+        return None
+
+    def label(self, label, node: "Teacher") -> None:
+        """
+        Apply a label to a known node.  Removes any exiting labels before adding the new one.
+        If the provided label is not valid UnknownLabel is raised.
+        If the provided node is not known UnknownNode is raised.
+        """
+        if label not in BUCKETS:
+            raise self.UnknownLabel(f'{label} is not a valid node label')
+        if self.__nodes.get(node.checksum_address):
+            self.unlabel(node=node)            # Remove any existing labels...
+            self.__marked[label].append(node)  # Set the new label
+        else:
+            raise self.UnknownNode(f'Cannot label an unknown node ({node}).')
+
+    def unlabel(self, node: "Teacher", label=None) -> None:
+        """Removes a label from a node, or if label is None, all labels are removed."""
+
+        # Remove one label
+        if label:
+            if node in self.__marked[label]:
+                self.__marked[label].remove(node)
+            return
 
-        if self._nodes.get(node):
-            del self._nodes[node]
+        # Remove all labels
+        node_labels = []
+        for _label in BUCKETS:
+            if node in self.__marked[_label]:
+                self.__marked[_label].remove(node)
+                node_labels.append(_label)
+
+        # extra check to ensure that nodes only ever have one label
+        if len(node_labels) > 1:
+            # well this is unexpected :(
+            self.log.warn(f"Unlabelled node ({node}), but it had multiple labels: {node_labels}")
+
+    def prune_bucket(self, label):
+        """Apply pruning strategies to a single node bucket once"""
+        try:
+            strategies = PRUNING_STRATEGIES[label]
+        except KeyError:
+            raise self.UnknownLabel(f'"{label}" is not a known node label.')
+        for node in self.get_nodes(label=label):
+            for strategy in strategies:
+                keep = strategy(node=node)
+                if not keep:
+                    del self.__nodes[node.checksum_address]  # prune node
+                    self.unlabel(node=node, label=label)     # prune corresponding label
+                    break  # this node is already doomed anyways
+                    # TODO: forget node from disk too
+                    # TODO: Trash can label?
+            else:
+                # Reinstate this node to good standing by un/relabeling
+                if label in PRUNING_STRATEGIES:
```

Comment:
For clarity, based on the operations done in this if-block, should we use BUCKETS here instead of PRUNING_STRATEGIES?

```suggestion
                if label in BUCKETS:
```

Reply:
Not every bucket has a strategy associated with it.

Reply:
👍 Right! ignore that suggestion.

Side note, why are we re-instating the node in this block? We are resetting the tracking if the node is determined to be kept for now and not yet pruned. Do we lose tracking data by resetting the node label tracking eg. max attempts?